### PR TITLE
Logs: collapse multi-line logs and expand with Log Details

### PR DESCRIPTION
**Pull request by kavindunilshan**
**What is this feature?**

This allows multi-line log-entries in Grafana Logs View to be collapsable and expandable just by clicking on that log. 

**Why do we need this feature?**

It enhances readability by selectively expanding relevant information like stack traces, streamlines navigation through logs, accelerates debugging processes, and overall improves the user experience by offering a more intuitive and efficient log browsing interface.

**Who is this feature for?**

 For developers and operators dealing with complex software systems.

**Which issue(s) does this PR fix?**:

The issue id [#74261](https://github.com/grafana/grafana/issues/74261) is fixed from this PR.

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/logs/components/LogRowMessage.test.tsx [hunk 1]
```diff
@@ -13,6 +13,7 @@ const setup = (propOverrides?: Partial<ComponentProps<typeof LogRowMessage>>, ro
   const styles = getLogRowStyles(theme);
   const props: ComponentProps<typeof LogRowMessage> = {
     wrapLogMessage: false,
+    isExpanded: false,
```
Comment (matyax) [inline public/app/features/logs/components/LogRowMessage.test.tsx:16]: If we add this feature we will need a test for these changes. I can help with that.

File: public/app/features/logs/components/LogRowMessage.tsx [hunk 2]
```diff
@@ -58,20 +59,38 @@ const LogMessage = ({ hasAnsi, entry, highlights, styles }: LogMessageProps) =>
   return <>{entry}</>;
 };
 
-const restructureLog = (line: string, prettifyLogMessage: boolean): string => {
+function extractFirstLine(input: string): string {
+  // Regular expression to match everything up to the first occurrence of '\r\n'
+  const regex = /^(.*?)(\r\n|$)/;
```
Comment (matyax) [inline public/app/features/logs/components/LogRowMessage.tsx:64]:
```suggestion
  const regex = /^(.*?)(\r\n|\n|$)/;
```
CRLF line endings are primarily used in Windows operating systems; Unix-based systems like Linux and macOS typically use LF for line endings.

File: public/app/features/logs/components/LogRowMessage.tsx [hunk 3]
```diff
@@ -58,20 +59,38 @@ const LogMessage = ({ hasAnsi, entry, highlights, styles }: LogMessageProps) =>
...
+  const match = input.match(regex);
+
+  if (match) {
+    return match[1];
+  } else {
+    return input;
+  }
+}
+
+const restructureLog = (isExpanded:boolean, line: string, prettifyLogMessage: boolean): string => {
```
Comment (matyax) [inline public/app/features/logs/components/LogRowMessage.tsx:74]:
```suggestion
const restructureLog = (line: string, prettifyLogMessage: boolean, isExpanded:boolean): string => {
```
As a good practice, data first, and options at the end.

**Review by matyax — CHANGES_REQUESTED** [review]
Hey @kavindunilshan this is great! Your contribution is much appreciated.

IMO, this change is relatively simple to make, but the potential of having a huge impact on a diverse number of users, each one with a peculiar type of log lines. This might greatly help a % of them, but harm the other.

For example, information loss by grabbing the first characters before the LF, now every log line is the same:


https://github.com/grafana/grafana/assets/1069378/356ee8d2-ea16-4edb-a737-8f82e52b7d4e

Will discuss and get back to you.

**Comment by matyax** [comment]
Once again, thanks a lot for this submission @kavindunilshan ! Unfortunately we decided not to move forward with this approach, for the reason I mentioned above, + the fact that we would need to add a control, somewhere, to allow users to enable or disable this new behavior.

If you're interested in following up and adressing this issue, our recommendation would be to look at https://github.com/grafana/grafana/issues/74262 and integrate it with `wrap lines`. The behavior of `Wrap lines` should be:
- Disabled: display the logs in a single line (maybe stripping out CRLF/LF characters)
- Enabled: lines are wrapped and new lines displayed.

**Comment by kavindunilshan** [comment]
Hey @matyax Thanks for the review! I completely understand your concerns about the potential impact on different log formats and the necessity for user control. 

While the wrap lines feature you mentioned is great, it's worth noting that since it applies to all the log lines, there are instances, such as when displaying error logs with stack traces, where expanding all logs at the same time might not be suitable.

I believe it would be more effective to implement a way to expand and collapse individual log lines as needed.

Regarding your point about needing to add a control for this behavior, I propose incorporating a separate selector, similar to the current wrap line selector option, to enable and disable this feature.

What do you think about this approach?

File: public/app/features/logs/components/LogRowMessage.tsx (hunk 3 above)
Comment (kavindunilshan) [reply public/app/features/logs/components/LogRowMessage.tsx:74]: Yes that way seems much better.

**Comment by matyax** [comment]
@kavindunilshan Expanding and collapsing individual log lines is not a possibility now. Instead, let's use `wrap lines` as a general toggle for showing single/multi lines, combined with displaying the full log line with new lines when you open log details. It would be a combination of this PR + https://github.com/grafana/grafana/issues/74262

If you're up for it, let's discuss it in this other follow up PR.

**Comment by matyax** [comment]
Hey @kavindunilshan ! I followed this UP in this PR: https://github.com/grafana/grafana/pull/88144 Feel free to chime in.

**Pull request closed without merging**